### PR TITLE
Fix condensed merging

### PR DESCRIPTION
**Pull request by pcm32**
This PR fixes a few issues encountered when testing this with Nancy manually:

- Condensed should be generated without headers
- Missing column definition in release
- doc tests for condensed merged

### CODE REVIEW

**Comment by anjaf** [comment]
I have tried running the merge_condensed_sdrfs.py script using the test data and the command line script but ran into a couple of errors. 

First, I tried with only the mandatory arguments. 
`merge_condensed_sdrfs.py -d ~/Atlasprod_software/MAGE-Tab-merger/test-data/condensed-merge -o ~/Desktop/test-condensed-merged -n E-TEST-1 -a E-CURD-31,E-GEOD-55866,E-GEOD-53197,E-GEOD-55482`
It gave me this error: 
```
Parsing E-CURD-31 condensed SDRF..
Parsing E-GEOD-55866 condensed SDRF..
Parsing E-GEOD-53197 condensed SDRF..
Complete list of covariates: 0 []
Number of connected components: 0
Traceback (most recent call last):
  File "/Users/anjaf/opt/miniconda3/envs/mtab-merger-test/bin/merge_condensed_sdrfs.py", line 162, in <module>
    conn_components = cluster_samples(cond=cond,
  File "/Users/anjaf/opt/miniconda3/envs/mtab-merger-test/bin/merge_condensed_sdrfs.py", line 133, in cluster_samples
    largest_conn_comp = max(conn_components, key=len)
ValueError: max() arg is an empty sequence
```

It seems to only works when specifying the co-variate as `-c "organism part"`. It could be the default value "organism_part" that isn't matching. Probably better use "organism part". 
This is something that should probably report a human-readable error to the user, when the co-variate isn't found in the available attribute categories.  

Now with adding `-c "organism part"`, I get a bit further but still fails:
`merge_condensed_sdrfs.py -d ~/Atlasprod_software/MAGE-Tab-merger/test-data/condensed-merge -o ~/Desktop/test-condensed-merged -n E-TEST-1 -a E-CURD-31,E-GEOD-55866,E-GEOD-53197,E-GEOD-55482 -c "organism part"`
```
Parsing E-CURD-31 condensed SDRF..
Parsing E-GEOD-55866 condensed SDRF..
Parsing E-GEOD-53197 condensed SDRF..
Parsing E-GEOD-55482 condensed SDRF..
Complete list of covariates: 4 ['root', 'silique', 'floral bud', 'aerial part']
Processing cov: root
Processing cov: silique
Processing cov: floral bud
Processing cov: aerial part
Number of connected components: 2
Largest connected component size: 3
CC number of batches: 3
CC batches: ['E-CURD-31', 'E-GEOD-53197', 'E-GEOD-55482']
CC number of covariates: 3
CC covariates: {'aerial part', 'root', 'floral bud'}

CC number of batches: 1
CC batches: ['E-GEOD-55866']
CC number of covariates: 1
CC covariates: {'silique'}

Parsing E-CURD-31 condensed SDRF..
Parsing E-GEOD-53197 condensed SDRF..
Parsing E-GEOD-55482 condensed SDRF..
Traceback (most recent call last):
  File "/Users/anjaf/opt/miniconda3/envs/mtab-merger-test/bin/merge_condensed_sdrfs.py", line 179, in <module>
    blessed_condensed.to_csv(path_or_buf=os.path.join(args.output, f"{args.new_accession}.condensed.sdrf.tsv"), sep="\t",
  File "/Users/anjaf/opt/miniconda3/envs/mtab-merger-test/lib/python3.8/site-packages/pandas/core/generic.py", line 3387, in to_csv
    return DataFrameRenderer(formatter).to_csv(
  File "/Users/anjaf/opt/miniconda3/envs/mtab-merger-test/lib/python3.8/site-packages/pandas/io/formats/format.py", line 1083, in to_csv
    csv_formatter.save()
  File "/Users/anjaf/opt/miniconda3/envs/mtab-merger-test/lib/python3.8/site-packages/pandas/io/formats/csvs.py", line 228, in save
    with get_handle(
  File "/Users/anjaf/opt/miniconda3/envs/mtab-merger-test/lib/python3.8/site-packages/pandas/io/common.py", line 642, in get_handle
    handle = open(
FileNotFoundError: [Errno 2] No such file or directory: '/Users/anjaf/Desktop/test-condensed-merged/E-TEST-1.condensed.sdrf.tsv'

The condensed file isn't being written. 
```

It seems that the output folder needs to exist already. It worked after I created the directory!
To improve I suggest to either have the program create the folder if it doesn't exit or add that in the documentation for the -o option.

File: merge_condensed_sdrfs.py [hunk 1]
```diff
@@ -75,6 +94,15 @@ def cluster_samples(cond: pd, main_covariate: str, covariate_type: str,
     :param batch_characteristic:
     :param batch_type:
     :return:
+    >>> accessions = ['E-GEOD-55866','E-GEOD-55482', 'E-CURD-31', 'E-GEOD-53197']
+    >>> input_path = get_test_fname("condensed-merge")
+    >>> cond_df = merged_condensed(accessions, input_path, batch_type='characteristic', batch_characteristic='study', new_accession="E-CURD-X") #doctest: +ELLIPSIS
+    Parsing ...
+    >>> conn_components = cluster_samples(cond=cond_df, main_covariate="organism part", covariate_type="characteristic", batch_characteristic="study", batch_type="characteristic", covariate_skip_values=[]) #doctest: +ELLIPSIS
```
Comment (anjaf) [inline merge_condensed_sdrfs.py:102]: Specifying "organism part" is critical here. The default from the command line argument "organism_part" doesn't work.

File: merge_condensed_sdrfs.py [hunk 2]
```diff
@@ -100,49 +128,73 @@ def cluster_samples(cond: pd, main_covariate: str, covariate_type: str,
...
+            covariates_cc = set()
+            for edge_cov in cc.edges.data('covariate'):
+                covariates_cc.update(edge_cov[2])
+            if not second_cc_covariates and first_cc_covariates:
+                second_cc_covariates.extend(covariates_cc)
+            if not first_cc_covariates:
+                first_cc_covariates.extend(covariates_cc)
+            print("CC number of covariates: {}".format(str(len(covariates_cc))))
+            print("CC covariates: {}".format(covariates_cc))
+            print()
```
Comment (anjaf) [inline merge_condensed_sdrfs.py]: I haven't seen the empty print statement used as a line break. `\n` in the format string is probably more explicit.

**Comment by pcm32** [comment]
Thanks for this great feedback @anjaf !

File: merge_condensed_sdrfs.py (hunk 1 above)
Comment (pcm32) [reply merge_condensed_sdrfs.py:102]: Changed the default now.

File: merge_condensed_sdrfs.py (hunk 2 above)
Comment (pcm32) [reply merge_condensed_sdrfs.py]: Resolved now!

**Comment by pcm32** [comment]
Thanks for the review @anjaf, will merge this as I have addressed your points, but let me know if you have any other comments to fix them.

**Review by anjaf — APPROVED** [review]

**Pull request merged**